src/Builder/BuilderCodec.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+namespace MongoDB\Builder;
+
+use LogicException;
+use MongoDB\Builder\Expression\Expression;
+use MongoDB\Builder\Expression\FieldPath;
+use MongoDB\Builder\Expression\Variable;
+use MongoDB\Builder\Query\OrQuery;
+use MongoDB\Builder\Stage\GroupStage;
+use MongoDB\Builder\Stage\ProjectStage;
+use MongoDB\Builder\Stage\Stage;
+use MongoDB\Codec\Codec;
+use MongoDB\Codec\DecodeIfSupported;
+use MongoDB\Codec\EncodeIfSupported;
+use MongoDB\Exception\UnsupportedValueException;
+use stdClass;
+
+use function array_is_list;
+use function array_merge;
+use function is_array;
+use function sprintf;
+
+class BuilderCodec implements Codec
+{
+    use DecodeIfSupported;
+    use EncodeIfSupported;
+
+    /** The first property is the operator value */
+    public const ENCODE_AS_SINGLE = 'single';
+
+    /** Arguments as encoded as a map of properties to their values */
+    public const ENCODE_AS_OBJECT = 'object';
+
+    /** Properties are encoded as a list of values, names are ignored */
+    public const ENCODE_AS_ARRAY = 'array';
+
+    public function canDecode($value): false
+    {
+        return false;
+    }
+
+    public function canEncode($value): bool
+    {
+        return $value instanceof Pipeline || $value instanceof Stage || $value instanceof Expression;
+    }
+
+    public function decode($value)
+    {
+        throw UnsupportedValueException::invalidDecodableValue($value);
+    }
+
+    public function encode($value): array|stdClass|string|int|float|bool|null
+    {
+        if (! $this->canEncode($value)) {
+            throw UnsupportedValueException::invalidEncodableValue($value);
+        }
+
+        // A pipeline is encoded as a list of stages
+        if ($value instanceof Pipeline) {
```

Comment:
This check and all subsequent checks should be handled by their own encoder; we can then leverage an encoder map to encode values.

Reply:
This is a draft implementation to get an overview of how operators and stages are dumped. I'll work on specificities on this tickets:

[PHPLIB-1256](https://jira.mongodb.org/browse/PHPLIB-1256)
[PHPLIB-1257](https://jira.mongodb.org/browse/PHPLIB-1257)
[PHPLIB-1258](https://jira.mongodb.org/browse/PHPLIB-1258)